### PR TITLE
Use different search algorithm in `Strings::contains_any`

### PR DESCRIPTION
**Pull request by Thomas1664**
Using `std::boyer_moore_horspool_searcher` results in approx. 3x speedup of `Strings::contains_any()` which is used by `contains_any_ignoring_c_comments()` during post build lint.

Note that this search algorithm is only useful when dealing with long strings: [cppreference](https://en.cppreference.com/w/cpp/utility/functional/boyer_moore_horspool_searcher)

### CODE REVIEW

**Comment by Thomas1664** [comment]
Interestingly, `std::boyer_moore_horspool_searcher` is not available on OSX.

File: src/vcpkg/base/strings.cpp [hunk 1]
```diff
@@ -473,7 +477,21 @@ bool Strings::contains_any_ignoring_hash_comments(StringView source, View<String
...
-    return Util::any_of(to_find, [=](StringView s) { return Strings::contains(source, s); });
+#ifdef __APPLE__
+    using std::experimental::boyer_moore_horspool_searcher;
+#else
+    using std::boyer_moore_horspool_searcher;
+#endif
+    for (const auto& subject : to_find)
+    {
+        auto found =
+            std::search(source.begin(), source.end(), boyer_moore_horspool_searcher(subject.begin(), subject.end()));
```
Comment (BillyONeal) [inline src/vcpkg/base/strings.cpp:488]: Creating the searcher should be hoist out of the loop

Comment (BillyONeal) [reply src/vcpkg/base/strings.cpp:488]: Nevermind I'm an idiot.

Comment (BillyONeal) [reply src/vcpkg/base/strings.cpp:488]: Although I do think this should be hoist to just after

```c++
auto files = fs.get_regular_files_recursive(dir, IgnoreErrors{});
```

in `postbuildlint.cpp`.

My concern is that `contains_any` looks like it's doing `contains`, not boyer_moore things, and I could easily see that being used in places boyer_moore would be inappropriate.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
I either think `contains_any` should be renamed to indicate that it only makes sense for long haystacks now, and/or it should be given a `View<boyer_moore_horspool_searcher>` instead of the raw text.

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks!

**Comment by autoantwort** [comment]
I now get 467 warnings of the form 
```
/Users/leanderSchulten/git_projekte/vcpkg-tool/include/vcpkg/base/strings.h:51: warning: 'boyer_moore_horspool_searcher<std::__wrap_iter<const char *>>' is deprecated: std::experimental::boyer_moore_horspool_searcher will be removed in LLVM 17. Use std::boyer_moore_horspool_searcher instead [-Wdeprecated-declarations]
In file included from /Users/leanderSchulten/git_projekte/vcpkg-tool/src/vcpkg/base/hash.cpp:6:
/Users/leanderSchulten/git_projekte/vcpkg-tool/include/vcpkg/base/strings.h:51:62: warning: 'boyer_moore_horspool_searcher<std::__wrap_iter<const char *>>' is deprecated: std::experimental::boyer_moore_horspool_searcher will be removed in LLVM 17. Use std::boyer_moore_horspool_searcher instead [-Wdeprecated-declarations]
    using boyer_moore_horspool_searcher = std::experimental::boyer_moore_horspool_searcher<std::string::const_iterator>;
                                                             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/experimental/functional:341:7: note: 'boyer_moore_horspool_searcher<std::__wrap_iter<const char *>>' has been explicitly marked deprecated here
class _LIBCPP_DEPRECATED_BOYER_MOORE_HORSPOOL_SEARCHER _LIBCPP_TEMPLATE_VIS boyer_moore_horspool_searcher {
      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/experimental/functional:90:60: note: expanded from macro '_LIBCPP_DEPRECATED_BOYER_MOORE_HORSPOOL_SEARCHER'
#  define _LIBCPP_DEPRECATED_BOYER_MOORE_HORSPOOL_SEARCHER _LIBCPP_DEPRECATED_("std::experimental::boyer_moore_horspool_searcher will be removed in LLVM 17. Use std::boyer_moore_horspool_searcher instead")
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__config:836:53: note: expanded from macro '_LIBCPP_DEPRECATED_'
#      define _LIBCPP_DEPRECATED_(m) __attribute__((deprecated(m)))
                                                    ^
```

**Comment by Thomas1664** [comment]
Honestly, I don't know what the best solution is. I could check for the feature test macro but then we don't get informed when we can remove this workaround.

**Pull request merged**